src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -678,15 +687,144 @@ vertexSet = Set.fromAscList . vertexList
+--
+-- @
+-- rightAdjacencyList 'empty'            == []
+-- rightAdjacencyList ('vertices' [] xs) == [ (x, []) | x <- 'nub' ('sort' xs) ]
+-- rightAdjacencyList ('vertices' xs []) == []
+-- rightAdjacencyList ('edge' x y)       == [(y, [x])]
+-- rightAdjacencyList ('star' x ys)      == [ (y, [x])  | y <- 'nub' ('sort' ys) ]
+-- @
+rightAdjacencyList :: AdjacencyMap a b -> [(b, [a])]
+rightAdjacencyList (BAM _ rl) = [ (v, Set.toAscList us) | (v, us) <- Map.toAscList rl ]
+
+-- | The sorted /adjacency list/ of the graph.
+-- Complexity: /O(n + m)/ time and /O(n)/ memory.
+--
+-- @
+-- adjacencyList 'empty'            == []
+-- adjacencyList ('vertices' xs ys) == [ (v, []) | v <- (map Left (sort (nub xs))) ++ (map Right (sort (nub ys))) ]
+-- adjacencyList ('edge' x y)       == [(Left x, [Right y]), (Right y, [Left x])]
+-- adjacencyList                  == 'Algebra.Graph.AdjacencyMap.adjacencyList' . 'fromBipartite'
+-- @
+adjacencyList :: AdjacencyMap a b -> [(Either a b, [Either a b])]
+adjacencyList g = [ (Left v, map Right us) | (v, us) <- leftAdjacencyList g ] ++
+                  [ (Right v, map Left us) | (v, us) <- rightAdjacencyList g ]
+
+-- | A list of alternating values of two types. The first type argument denotes
+-- the type of the head.
+--
+-- With @OverloadedLists@ extension it is possible to use standard list
+-- notation to make a 'List' of two coincidential types, e.g.
+-- 
+-- @
+-- [1, 2, 3, 4, 5] :: List Int Int
+-- @
+--
+-- This property is heavily used in the examples below.
+--
+-- The 'Show' instance matches the 'Show' instance for lists.
+--
+-- @
+-- 'show' Empty                              == "[]"
+-- 'show' ([1, 2, 3] :: List Int Int)        == "[1,2,3]"
+-- 'show' (Cons 1 (Cons "a" (Cons 3 Empty))) == "[1,\\"a\\",3]"
+-- @
+data List a b = Empty | Cons a (List b a)
+     deriving (Eq, Generic)
+
+instance (Show a, Show b) => Show (List a b) where
+    showsPrec _ = showListWith dropEither . toListEither
+
+toListEither :: List a b -> [Either a b]
+toListEither = go . Left
+    where
+        go :: Either (List a b) (List b a) -> [Either a b]
+        go (Left  Empty)       = []
+        go (Right Empty)       = []
+        go (Left  (Cons x xt)) = Left  x : go (Right xt)
+        go (Right (Cons x xt)) = Right x : go (Left  xt)
+
+dropEither :: (Show a, Show b) => Either a b -> ShowS
+dropEither (Left x)  = showsPrec 10 x
+dropEither (Right x) = showsPrec 10 x
+
+instance IsList (List a a) where
+    type Item (List a a) = a
+
+    fromList []     = Empty
+    fromList (x:xt) = Cons x (fromList xt)
+
+    toList Empty       = []
+    toList (Cons x xt) = x:(toList xt)
+
+-- | Construct a 'List' of even length from a list of pairs.
+--
+-- @
+-- fromEvenList []                   == 'Empty'
+-- fromEvenList [(1, 2), (3, 4)]     == [1, 2, 3, 4] :: 'List' Int Int
+-- fromEvenList [(1, "a"), (2, "b")] == 'Cons' 1 ('Cons' "a" ('Cons' 2 ('Cons' "b" 'Empty')))
+-- @
+fromEvenList :: [(a, b)] -> List a b
```

Comment:
It may be better to name these functions `toListEven` and `toListOdd` so that can be found when looking for "toList".